devicetypes/stelpro/stelpro-ki-thermostat.src/stelpro-ki-thermostat.groovy

```diff
@@ -0,0 +1,515 @@
+	}
+}
+
+def parse(String description) {
+	// If the user installed with an old DTH version, update so that the new mobile client will work
+	if (!device.currentValue("supportedThermostatModes")) {
+		configureSupportedRanges()
+	}
+
+	if (description == "updated") {
+		return null
+	}
+
+	// Class, version
+	def map = createEvent(zwaveEvent(zwave.parse(description, [0x40:2, 0x43:2, 0x31:3, 0x42:1, 0x20:1, 0x85: 2])))
+	if (!map) {
+		return null
+	}
+
+	def result = [map]
+	// This logic is to appease the (now deprecated but still sort-of used) consolidated
+	// Thermostat capability gods.
+	if (map.isStateChange && map.name == "heatingSetpoint") {
+		result << createEvent([
+				name: "thermostatSetpoint",
+				value: map.value,
+				unit: map.unit,
+				data: [thermostatSetpointRange: thermostatSetpointRange]
+			])
+	}
+
+	log.debug "Parse returned $result"
+	result
+}
+
+def updateWeather() {
+	log.debug "updating weather"
+	def weather
+	// If there is a zipcode defined, weather forecast will be sent. Otherwise, no weather forecast.
+	if (settings.zipcode) {
+		log.debug "ZipCode: ${settings.zipcode}"
+		weather = getWeatherFeature("conditions", settings.zipcode)
+
+		// Check if the variable is populated, otherwise return.
+		if (!weather) {
+			log.debug("Something went wrong, no data found.")
+			return false
+		}
+
+		def locationScale = getTemperatureScale()
+		def tempToSend = (locationScale == "C") ? weather.current_observation.temp_c : weather.current_observation.temp_f
+		log.debug("Outdoor Temperature: ${tempToSend} ${locationScale}")
+		// Right now this can disrupt device health if the device is
+		// currently offline -- it would be erroneously marked online.
+		//sendEvent(name: 'outsideTemp', value: tempToSend)
+		setOutdoorTemperature(tempToSend)
+	}
+}
+
+def scheduledUpdateWeather() {
+	def actions = updateWeather()
+
+	if (actions) {
+		sendHubCommand(actions)
+	}
+}
+
+// Command Implementations
+
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ **/
+def ping() {
+	log.debug "ping()"
+	zwave.sensorMultilevelV3.sensorMultilevelGet().format()
+}
+
+def poll() {
+	log.debug "poll()"
+	delayBetween([
+			updateWeather(),
+			zwave.thermostatOperatingStateV1.thermostatOperatingStateGet().format(),
+			zwave.thermostatModeV2.thermostatModeGet().format(),
+			zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1).format(),
+			zwave.sensorMultilevelV3.sensorMultilevelGet().format() // current temperature
+		], 100)
+}
+
+// Event Generation
+def zwaveEvent(thermostatsetpointv2.ThermostatSetpointReport cmd) {
+	def cmdScale = cmd.scale == 1 ? "F" : "C"
+	def temp;
+	float tempfloat;
+	def map = [:]
+
+	if (cmd.scaledValue >= 327 ||
+		cmd.setpointType != thermostatsetpointv2.ThermostatSetpointReport.SETPOINT_TYPE_HEATING_1) {
+		return [:]
+	}
+	temp = convertTemperatureIfNeeded(cmd.scaledValue, cmdScale, cmd.precision)
+	tempfloat = (Math.round(temp.toFloat() * 2)) / 2
```

Comment:
just point out that this is super weird

Reply:
If I remember well, this was done to ensure that the displayed temperature was rounded properly and was using the proper temperature format